### PR TITLE
leveldb: fix write pause state

### LINKED ISSUE #242: WritePaused:true continued for long time even L0 files number below WriteL0PauseTrigger
I print the DBStats every 5 second

set WriteL0PauseTrigger = 24, when the L0 file number reach 24, it trigger write pause, after a compaction, the L0 file number become 16, but `WritePaused:true` continued for long time after seems another compaction the L0 file number reach 0

goleveldb git commit version: c4c61651e9e3
```
LevelSizes:[588599075 4292188559 474488269] LevelTablesCounts:[24 64 7] LevelRead:[0 52202337026 0] LevelWrite:[5729149421 51828463508 0] LevelDurations:[4m34.041881282s 1h36m26.386456317s 0s]}^[[0m
2018/11/02 16:37:53 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:false AliveSnapshots:0 AliveIterators:0 IOWrite:75726140018 IORead:56320542805 BlockCacheSize:1136993 OpenedTablesCount:70 LevelSizes:[588599075 4292188559 474488269] LevelTablesCounts:[24 64 7] LevelRead:[0 52202337026 0] LevelWrite:[5729149421 51828463508 0] LevelDurations:[4m34.041881282s 1h36m26.386456317s 0s]}^[[0m
2018/11/02 16:37:58 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:75777037633 IORead:56365165996 BlockCacheSize:1136993 OpenedTablesCount:71 LevelSizes:[588599075 4292188559 474488269] LevelTablesCounts:[24 64 7] LevelRead:[0 52202337026 0] LevelWrite:[5729149421 51828463508 0] LevelDurations:[4m34.041881282s 1h36m26.386456317s 0s]}^[[0m
2018/11/02 16:38:03 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:75823123358 IORead:56411251720 BlockCacheSize:1136993 OpenedTablesCount:71 LevelSizes:[588599075 4292188559 474488269] LevelTablesCounts:[24 64 7] LevelRead:[0 52202337026 0] LevelWrite:[5729149421 51828463508 0] LevelDurations:[4m34.041881282s 1h36m26.386456317s 0s]}^[[0m
2018/11/02 16:38:08 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:75868668638 IORead:56456797013 BlockCacheSize:1136993 OpenedTablesCount:72 LevelSizes:[588599075 4292188559 474488269] LevelTablesCounts:[24 64 7] LevelRead:[0 52202337026 0] LevelWrite:[5729149421 51828463508 0] LevelDurations:[4m34.041881282s 1h36m26.386456317s 0s]}^[[0m
2018/11/02 16:38:13 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:75921921608 IORead:56509595534 BlockCacheSize:1136993 OpenedTablesCount:73 LevelSizes:[588599075 4292188559 474488269] LevelTablesCounts:[24 64 7] LevelRead:[0 52202337026 0] LevelWrite:[5729149421 51828463508 0] LevelDurations:[4m34.041881282s 1h36m26.386456317s 0s]}^[[0m
2018/11/02 16:38:18 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:75986163044 IORead:56573370792 BlockCacheSize:1136993 OpenedTablesCount:73 LevelSizes:[588599075 4292188559 474488269] LevelTablesCounts:[24 64 7] LevelRead:[0 52202337026 0] LevelWrite:[5729149421 51828463508 0] LevelDurations:[4m34.041881282s 1h36m26.386456317s 0s]}^[[0m
2018/11/02 16:38:23 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76061346007 IORead:56648144115 BlockCacheSize:1136993 OpenedTablesCount:73 LevelSizes:[588599075 4292188559 474488269] LevelTablesCounts:[24 64 7] LevelRead:[0 52202337026 0] LevelWrite:[5729149421 51828463508 0] LevelDurations:[4m34.041881282s 1h36m26.386456317s 0s]}^[[0m
2018/11/02 16:38:28 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76109216117 IORead:56713897690 BlockCacheSize:1136993 OpenedTablesCount:18 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:38:33 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76148912490 IORead:56759314298 BlockCacheSize:1136993 OpenedTablesCount:18 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:38:38 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76195528673 IORead:56805930442 BlockCacheSize:1136993 OpenedTablesCount:19 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:38:43 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76240745984 IORead:56851147739 BlockCacheSize:1136993 OpenedTablesCount:19 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:38:48 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76285704218 IORead:56896106251 BlockCacheSize:1136993 OpenedTablesCount:20 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:38:53 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76330619505 IORead:56941021399 BlockCacheSize:1136993 OpenedTablesCount:21 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:38:58 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76376885543 IORead:56987287431 BlockCacheSize:1136993 OpenedTablesCount:21 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:39:03 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76423241009 IORead:57033642335 BlockCacheSize:1136993 OpenedTablesCount:22 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:39:08 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76468760814 IORead:57079163612 BlockCacheSize:1136993 OpenedTablesCount:23 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:39:13 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:76513902874 IORead:57124305686 BlockCacheSize:1136993 OpenedTablesCount:23 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
```

```
LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:47:03 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:80824091365 IORead:61432551878 BlockCacheSize:1136993 OpenedTablesCount:84 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:47:08 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:764481 WriteDelayDuration:38m32.794866532s WritePaused:true AliveSnapshots:0 AliveIterators:0 IOWrite:80898808435 IORead:61506846419 BlockCacheSize:1136993 OpenedTablesCount:84 LevelSizes:[393367132 4474016654 474488269] LevelTablesCounts:[16 67 7] LevelRead:[0 56689757528 0] LevelWrite:[5729149421 56302480162 0] LevelDurations:[4m34.041881282s 1h44m56.273628238s 0s]}^[[0m
2018/11/02 16:47:13 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:930308 WriteDelayDuration:50m49.80274911s WritePaused:false AliveSnapshots:0 AliveIterators:0 IOWrite:80974243946 IORead:61574314806 BlockCacheSize:1136993 OpenedTablesCount:2 LevelSizes:[0 4846451499 474488269] LevelTablesCounts:[0 72 7] LevelRead:[0 61557141314 0] LevelWrite:[5729149421 61148931661 0] LevelDurations:[4m34.041881282s 1h53m41.040576742s 0s]}^[[0m
2018/11/02 16:47:18 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:930308 WriteDelayDuration:50m49.80274911s WritePaused:false AliveSnapshots:0 AliveIterators:0 IOWrite:81077816494 IORead:61631333009 BlockCacheSize:1136993 OpenedTablesCount:3 LevelSizes:[0 4846451499 474488269] LevelTablesCounts:[0 72 7] LevelRead:[0 61557141314 0] LevelWrite:[5729149421 61148931661 0] LevelDurations:[4m34.041881282s 1h53m41.040576742s 0s]}^[[0m
2018/11/02 16:47:23 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:930308 WriteDelayDuration:50m49.80274911s WritePaused:false AliveSnapshots:0 AliveIterators:0 IOWrite:81159715059 IORead:61688433409 BlockCacheSize:1136993 OpenedTablesCount:3 LevelSizes:[25093694 4846451499 474488269] LevelTablesCounts:[1 72 7] LevelRead:[0 61557141314 0] LevelWrite:[5754243115 61148931661 0] LevelDurations:[4m35.262208936s 1h53m41.040576742s 0s]}^[[0m
2018/11/02 16:47:28 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:930308 WriteDelayDuration:50m49.80274911s WritePaused:false AliveSnapshots:0 AliveIterators:0 IOWrite:81237967407 IORead:61754859619 BlockCacheSize:1136993 OpenedTablesCount:4 LevelSizes:[25093694 4846451499 474488269] LevelTablesCounts:[1 72 7] LevelRead:[0 61557141314 0] LevelWrite:[5754243115 61148931661 0] LevelDurations:[4m35.262208936s 1h53m41.040576742s 0s]}^[[0m
2018/11/02 16:47:33 storage.go:359: ^[[0;37m[info] DBStats: {WriteDelayCount:930308 WriteDelayDuration:50m49.80274911s WritePaused:false AliveSnapshots:0 AliveIterators:0 IOWrite:81323574305 IORead:61803945277 BlockCacheSize:1136993 OpenedTablesCount:5 LevelSizes:[49899592 4846451499 474488269] LevelTablesCounts:[2 72 7] LevelRead:[0 61557141314 0] LevelWrite:[5779049013 61148931661 0] LevelDurations:[4m36.487803469s 1h53m41.040576742s 0s]}^[[0m
```

### PR DESCRIPTION
**Pull request by rjl493456442**
Fix #242

### CODE REVIEW

File: leveldb/db_compaction.go [hunk 1]
```diff
@@ -844,6 +844,11 @@ func (db *DB) tCompaction() {
 			switch cmd := x.(type) {
 			case cAuto:
 				if cmd.ackC != nil {
+					// Check the write pause state before caching it.
+					if db.resumeWrite() {
+						x.ack(nil)
+						continue
```
Comment (syndtr) [inline leveldb/db_compaction.go]: Don't use `continue` here, use `if else` instead. If compaction need to happen then let it happen. Thanks!

Comment (rjl493456442) [reply leveldb/db_compaction.go]: Updated.

**Comment by syndtr** [comment]
Merged. Thanks!

**Comment by july2993** [comment]
how's this test?
is there such contention case
- found L0 num >= pause num and trigger write pause
- compacting L0 and continue to compact L > 0
- send a cmd to `tcompCmdC`
- need to wait the not relate compaction finished to resume write

**Pull request merged**